univariate-portfolio-sorts.qmd

```diff
@@ -102,27 +102,34 @@ The results indicate that we cannot reject the null hypothesis of average return
 
 ## Functional Programming for Portfolio Sorts
 
-Now we take portfolio sorts to the next level. We want to be able to sort stocks into an arbitrary number of portfolios. For this case, functional programming is very handy: we employ the [curly-curly](https://www.tidyverse.org/blog/2019/06/rlang-0-4-0/#a-simpler-interpolation-pattern-with-)-operator to give us flexibility concerning which variable to use for the sorting, denoted by `var`.\index{Curly-curly} We use `quantile()` to compute breakpoints for `n_portfolios`. Then, we assign portfolios to stocks using the `findInterval()` function. The output of the following function is a new column that contains the number of the portfolio to which a stock belongs.\index{Functional programming}
+Now we take portfolio sorts to the next level. We want to be able to sort stocks into an arbitrary number of portfolios. For this case, functional programming is very handy: we employ `all_of` inside a `pull()` to give us flexibility concerning which variable to use for the sorting, denoted by `sorting_variable`. We use `quantile()` to compute breakpoints for `n_portfolios`. Then, we assign portfolios to stocks using the `findInterval()` function. The output of the following function is a new column that contains the number of the portfolio to which a stock belongs.\index{Functional programming} 
+
+In some applications, the variable used for the sorting might be clustered (e.g., at a lower bound of 0). Then, multiple breakpoints may be identical, leading to empty portfolios. This can be easily identified in the output by checking if the number of the portfolios' constituents is constant across portfolios. Similarly, some portfolios have a very small number of stocks, e.g., at the beginning of the sample.
 
 ```{r}
-assign_portfolio <- function(data, var, n_portfolios) {
+assign_portfolio <- function(data, 
+                             sorting_variable, 
+                             n_portfolios) {
+  # Compute breakpoints
   breakpoints <- data |>
-    reframe(
-      breakpoint = quantile(
-        {{ var }},
-        probs = seq(0, 1, length.out = n_portfolios + 1),
-        na.rm = TRUE)
-      ) |>
-    pull(breakpoint) |>
-    as.numeric()
+    pull(all_of(sorting_variable)) |>
```

Comment:
Würde lieber curly-curly behalten. Geht nicht auch pull({{sorting_variable}})?

Reply:
curly-curly does work (of course, the input has to be changed as well). I found all_of() simpler to understand though, but I guess some learning might be good. @voigtstefan What do you think?

Reply:
Ja, in der selection ist all_of() schon praktisch, aber wir haben curly-curly eingeführt, um zu zeigen, was man damit machen kann. all_of() ersetzt sicher nicht curly-curly in allen use cases. Daher würde ich curly-curly behalten.

Reply:
Reintroduced the curly-curly operator in Commit https://github.com/tidy-finance/website/commit/a24d695f1a6e44a194c3fc0a6fc58882eec789b9. Should I also add this to the changelog?

Reply:
What do you mean by "this"? The change in assign_portfolio, yes. The reitroduction of curly-curly, no.

Reply:
this = the commit https://github.com/tidy-finance/website/commit/a24d695f1a6e44a194c3fc0a6fc58882eec789b9. Or should I just mention the original commit?

Reply:
I'd just link this pull request and briefly summarize what has happened.

Reply:
- [May 13, 2023, Commit 1232646:](https://github.com/tidy-finance/website/commit/1232646023dfd8b551d9d047d40c310550564b6b) We changed the assign_portfolio()-functions in Chapters [Univariate Portfolio Sorts](https://www.tidy-finance.org/univariate-portfolio-sorts.html), [Size Sorts and p-Hacking](https://www.tidy-finance.org/size-sorts-and-p-hacking.html), [Value and Bivariate Sorts](https://www.tidy-finance.org/value-and-bivariate-sorts.html), and [Replicating Fama and French Factors](https://www.tidy-finance.org/replicating-fama-and-french-factors.html). Additionally, we added a small explanation to potential issues with the function for clustered sorting variables in Chaptes [Univariate Portfolio Sorts](https://www.tidy-finance.org/univariate-portfolio-sorts.html) (see [Commit 24f9e60](https://github.com/tidy-finance/website/commit/24f9e6048f4f23a37e508fd3b98a8d73854e0dad)).